### PR TITLE
Change MaxUnpool to accept tensors with 0-dim batch sizes.

### PR DESCRIPTION
**Pull request by v0dro**
Part of the fix for https://github.com/pytorch/pytorch/issues/38115.

Changes the `MaxUnpool` module to work with 0-dimensions batch sizes.

### CODE REVIEW

File: test/test_nn.py [hunk 1]
```diff
@@ -13570,6 +13570,34 @@ def test_MaxPool_zero_batch_dim(self, device):
...
             mod(inp)
 
+    @onlyOnCPUAndCUDA
+    def test_MaxUnpool_zero_batch_dim(self, device):
+        pool = torch.nn.MaxPool1d(2, stride=2, return_indices=True).to(device)
+        unpool = torch.nn.MaxUnpool1d(2, stride=2).to(device)
+        inp = torch.randn(0, 10, 10, requires_grad=True, device=device)
+        output, indices = pool(inp)
+        output.requires_grad_(True)
+        unpool(output, indices).sum().backward()
```
Comment (jbschlosser) [inline test/test_nn.py:13580]: Should there be a check that grads are zeros as well? Same throughout

**Review by jbschlosser — COMMENTED** [review]
Thanks for the update! Implementation looks fine; one minor testing Q below

File: test/test_nn.py (hunk 1 above)
Comment (v0dro) [reply test/test_nn.py]: Done.

**Comment by v0dro** [comment]
I'm unable to see the results of the failing test. Any idea why?

**Comment by jbschlosser** [comment]
@v0dro Neither can I; I think it's infra failure

File: test/test_nn.py [hunk 2]
```diff
@@ -13783,6 +13783,37 @@ def test_MaxPool_zero_batch_dim(self, device):
...
+    def test_MaxUnpool_zero_batch_dim(self, device):
+        pool = torch.nn.MaxPool1d(2, stride=2, return_indices=True).to(device)
+        unpool = torch.nn.MaxUnpool1d(2, stride=2).to(device)
+        inp = torch.randn(0, 10, 10, requires_grad=True, device=device)
+        output, indices = pool(inp)
+        output.requires_grad_(True)
+        unpool_out = unpool(output, indices)
+        unpool_out.sum().backward()
+
+        self.assertEqual(inp.grad, torch.zeros_like(inp))
```
Comment (jbschlosser) [inline test/test_nn.py:13796]: I think both grads and outputs should be checked here and below.

Also, probably outside of the scope of this PR, but it seems gradgrad checks aren't being done and that led to https://github.com/pytorch/pytorch/issues/64306#issuecomment-910369537. At some point, we should add testing for this

Comment (v0dro) [reply test/test_nn.py:13796]: OK thanks. Will add the tests and update.

Comment (v0dro) [reply test/test_nn.py:13796]: On a side note, I'm checking for the same things as the `_test_module_empty_input` function is checking (it does not check the outputs). Do you want to add output checks to the that function too?

Comment (jbschlosser) [reply test/test_nn.py:13776]: Yes please :)

Comment (jbschlosser) [reply test/test_nn.py:13776]: Couldn't `_test_module_empty_input` be invoked here instead of replicating the checks? If not, maybe it could be refactored slightly. Ideally it'd be nice if it was flexible enough to be used consistently throughout the zero batch size tests

**Review by jbschlosser — APPROVED** [review]
LGTM, testing for gradgrad can be bumped to a future PR

File: test/test_nn.py (hunk 2 above)
Comment (v0dro) [reply test/test_nn.py:13776]: Yeah true that. Should I make an issue and add it as a refactoring step later?
